core/src/main/java/org/keycloak/Feature.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak;
+
+import org.keycloak.common.Version;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Properties;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * @author <a href="mailto:mstrukel@redhat.com">Marko Strukelj</a>
+ */
+public enum Feature {
+
+    /**
+     *
+     *   When you add, remove, or change a feature, you also have to modify methods: fromCaption() and isEnabledByDefault()
```

Comment:
I think steps on adding/removing features should be added to a readme in misc rather than documented here. Then we can just reference that file in here.

Also, I'm a bit concerned with the amount of places that needs to be updated. Is there something we can do to reduce this?

Reply:
Sure, but it's important that when making additions one does not forget to look at those places. Hopefully I added enough tests to make sure most of those locations aren't overlooked.

Reply:
Regarding all the places involved ... I'm not happy with that either but it has to do with how packaging is done, and with a decision to explicitly spell out the subsystem changes in .cli files. With regards to .cli files there are three options - we include all keycloak-server subsystem changes in them (ALL), we don't include any keycloak-server subsystem changes in them (NONE), or we include some keycloak-server subsystem changes in them (SOME). We've been practising option ALL. Effectively we could be practicing option NONE if we performed upgrades inside keycloak-server. One problem is that in order to ensure that model change is saved in standalone.xml file we would need to increment subsystem version every time we make changes. Without it it would, nevertheless, be reinitialised with defaults upon every start, and it would be visible to jboss-cli and runtime DMR config, just not visible in standalone.xml - but that is bad as users can't easily open the file and see available features, system properties through which they are controlled, and tweak it easily. Incrementing version also means changing all the docs with examples of subsystem xml. I think it's better to create another issue for that then. Option SOME would be to omit adding features to .cli files with same effect of them not being visible in standalone.xml, but we would not remove any existing keycloak-server subsystem entries from .cli files and add logic for their initialisation into the subsystem. 

For other files, we should go over each one, and consider if anything can be done to remove it. The danger is that while we might devise a plan how to have things in one place it would come with horrible scripting extras at maven plugin level to use and transform the same content in various places, which makes packaging hard to understand, hard to debug, hard to change.

Reply:
Let's add file in misc, but also refer to it in the Javadoc in Features.java